dali/test/python/experimental_mode/test_batch.py

```diff
@@ -0,0 +1,149 @@
+@params(("cpu",), ("cuda",))
+def test_batch_construction_with_torch_tensor(device_type):
+    import torch
+
+    data = torch.tensor([[1, 2, 3], [4, 5, 6]], device=device_type, dtype=torch.int32)
+    b = D.as_batch(data)
+    assert b.device == D.Device("gpu" if device_type == "cuda" else device_type)
+    assert b.dtype == D.int32
+    assert b.batch_size == 2
+    assert b.ndim == 1
+    assert b.shape == [(3,), (3,)]
+    assert b.layout is None
+    ref = torch.from_dlpack(D.as_tensor(b)._backend)
+    assert torch.equal(data[0], ref[0])
+    assert torch.equal(data[1], ref[1])
+
+
+@params(("cpu",), ("gpu",))
+def test_batch_construction_with_tensor(device_type):
+    t = D.tensor(np.array([1, 2, 3], dtype=np.uint8), device=device_type, layout="X")
+    b = D.Batch([t])
+    assert b.device == D.Device(device_type)
+    assert b.dtype == D.uint8
+    assert b.layout == "X"
+    assert b.batch_size == 1
+    assert b.ndim == 1
+    assert b.shape == [(3,)]
+
+
+@params(("cpu",), ("gpu",))
+def test_batch_construction_with_conversion(device_type):
+    data = [np.float64([1]), np.float64([1.25, 2.2, 0x1000001])]
+    data_i32 = [np.int32([1]), np.int32([1, 2, 0x1000001])]
+    data_fp32 = [np.float32([1]), np.float32([1.25, 2.2, 0x1000000])]
+    # loss of precision --------------------------^
+    orig = D.Batch(data, device=device_type).evaluate()
+    # convert from a list of tensors
+    i32 = D.Batch(data, device=device_type, dtype=D.int32).evaluate()
+    # convert from a TensorList object
+    fp32 = D.Batch(orig._backend, device=device_type, dtype=D.float32).evaluate()
+    assert orig.dtype == D.float64
+    assert orig.device == D.Device(device_type)
+    assert orig._backend.dtype == D.float64.type_id
+    assert i32.dtype == D.int32
+    assert i32.device == D.Device(device_type)
+    assert i32._backend.dtype == D.int32.type_id
+    assert fp32.dtype == D.float32
+    assert fp32.device == D.Device(device_type)
+    assert fp32._backend.dtype == D.float32.type_id
+    assert batch_equal(data, asnumpy(orig))
+    assert batch_equal(data_i32, asnumpy(i32))
+    assert batch_equal(data_fp32, asnumpy(fp32))
+
+
+@params(("cpu",), ("gpu",))
+def test_batch_properties_from_tensor(device_type):
+    t = D.tensor(np.array([1, 2, 3], dtype=np.uint8), device=device_type, layout="X")
+    b = D.Batch([t])
+    assert b.device == D.Device(device_type)
+    assert b.dtype == D.uint8
+    assert b.layout == "X"
+    assert b.batch_size == 1
+    assert b.ndim == 1
+    assert b.shape == [(3,)]
+
+
+@params(("cpu",), ("gpu",))
+def test_batch_properties_clone(device_type):
+    t = D.tensor(np.array([1, 2, 3], dtype=np.uint8))
+    src = D.Batch([t], device=device_type, layout="X")
+    b = D.batch(src)
+    assert b.device == D.Device(device_type)
+    assert b.dtype == D.uint8
+    assert b.layout == "X"
+    assert b.batch_size == 1
+    assert b.ndim == 1
+    assert b.shape == [(3,)]
+
+
+def test_batch_subscript_per_sample():
```

Comment:
This PR introduces very complex slicing logic, allowing indexing, slicing and Ellipsis, which is not tested